src/algebra/homology/chain_complex.lean

```diff
@@ -0,0 +1,128 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.graded_object
+import category_theory.differential_object
+
+/-!
+# Chain complexes
+
+We define a chain complex in `V` as a differential `ℤ`-graded object in `V`.
+
+This is fancy language for the obvious definition,
+and it seems we can use it straightforwardly:
+
+```
+example (C : chain_complex V) : C.X 5 ⟶ C.X 6 := C.d 5
+```
+
+We define the forgetful functor to `ℤ`-graded objects, and show that
+`chain_complex V` is concrete when `V` is, and `V` has coproducts.
+-/
+
+universes v u
+
+open category_theory
+open category_theory.limits
+
+section
+variables (V : Type u) [𝒱 : category.{v} V]
+include 𝒱
+
+variables [has_zero_morphisms.{v} V]
+
+/--
+A chain complex in `V` is "just" a differential `ℤ`-graded object in `V`.
+-/
+-- For now the "shift" is fixed to be the +1 direction,
+-- making this a cochain complex, rather than a chain complex.
+def chain_complex : Type (max v u) :=
+differential_object.{v} (graded_object ℤ V)
+
+-- The chain groups of a chain complex `C` are accessed as `C.X i`,
+-- and the differentials as `C.d i : C.X i ⟶ C.X (i+1)`.
+example (C : chain_complex V) : C.X 5 ⟶ C.X 6 := C.d 5
+
+variables {V}
+/--
+A convenience lemma for morphisms of differential graded objects,
+picking out one component of the commutation relation.
+-/
+-- Could this be a simp lemma? Which way?
+lemma category_theory.differential_object.hom.comm_at {X Y : differential_object.{v} (graded_object ℤ V)}
+  (f : X ⟶ Y) (i : ℤ) : f.f i ≫ Y.d i = X.d i ≫ f.f (i+1) := congr_fun f.comm i
+end
+
+namespace chain_complex
+variables {V : Type u} [𝒱 : category.{v} V]
+include 𝒱
+
+variables [has_zero_morphisms.{v} V]
+
+@[simp]
+lemma d_squared (C : chain_complex.{v} V) (i : ℤ) :
+  C.d i ≫ C.d (i+1) = 0 :=
+congr_fun (C.d_squared) i
+
+variables (V)
+
+instance category_of_chain_complexes : category.{v} (chain_complex V) :=
+by { dsimp [chain_complex], apply_instance }.
+
+-- The components of a chain map `f : C ⟶ D` are accessed as `f.f i`.
+example {C D : chain_complex V} (f : C ⟶ D) : C.X 5 ⟶ D.X 5 := f.f 5
+example {C D : chain_complex V} (f : C ⟶ D) : f.f ≫ D.d = C.d ≫ f.f[[1]] := f.comm
+example {C D : chain_complex V} (f : C ⟶ D) : f.f 5 ≫ D.d 5 = C.d 5 ≫ f.f 6 := congr_fun f.comm 5
+
+/-- The forgetful functor from chain complexes to graded objects, forgetting the differential. -/
+def forget : (chain_complex V) ⥤ (graded_object ℤ V) :=
+differential_object.forget _
+
+instance forget_faithful : faithful (forget V) :=
+differential_object.forget_faithful _
+
+instance has_zero_morphisms : has_zero_morphisms.{v} (chain_complex V) :=
+by { dsimp [chain_complex], apply_instance }.
+
+variables [has_zero_object.{v} V]
+
+instance has_zero_object : has_zero_object.{v} (chain_complex V) :=
+by { dsimp [chain_complex], apply_instance, }
```

Comment:
Just to be consistent with the `has_zero_morphisms` instance above ;)
```suggestion
by { dsimp [chain_complex], apply_instance }
```

Reply:
I'll make it consistent the other way. :-) I like the trailing commas, so when you put the cursor inside the `}` you get that satisfying "no goals".